src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/RestHealthCheckService.cs

```diff
@@ -1,22 +1,110 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
 
 namespace Microsoft.Azure.SignalR.Management
 {
     internal class RestHealthCheckService : IHostedService
     {
-        public Task StartAsync(CancellationToken cancellationToken)
+        //internal by test
+        //An acceptable time to wait before retry when clients negotiate fail
+        internal static readonly TimeSpan CheckInterval = TimeSpan.FromSeconds(10);
+        internal static readonly TimeSpan RetryInterval = TimeSpan.FromSeconds(1);
+        private const int MaxRetries = 2;
+
+        private readonly RestClientFactory _clientFactory;
+        private readonly IServiceEndpointManager _serviceEndpointManager;
+        private readonly ILogger<RestHealthCheckService> _logger;
+        private readonly string _hubName;
+
+        private readonly TimerAwaitable _timer = new(CheckInterval, CheckInterval);
+
+        public RestHealthCheckService(RestClientFactory clientFactory, IServiceEndpointManager serviceEndpointManager, ILogger<RestHealthCheckService> logger, string hubName)
         {
-            return Task.CompletedTask;
+            _clientFactory = clientFactory;
+            _serviceEndpointManager = serviceEndpointManager;
+            _logger = logger;
+            _hubName = hubName;
         }
 
-        public Task StopAsync(CancellationToken cancellationToken)
+        public async Task StartAsync(CancellationToken cancellationToken)
         {
+            // wait for the first health check finished
+            await CheckEndpointHealthAsync();
+            _ = LoopAsync();
+        }
+
+        public Task StopAsync(CancellationToken _)
+        {
+            _timer.Stop();
             return Task.CompletedTask;
         }
+
+        private async Task CheckEndpointHealthAsync()
+        {
+            await Task.WhenAll(_serviceEndpointManager.GetEndpoints(_hubName).Select(async endpoint =>
+            {
+                var retry = 0;
+                var isHealthy = false;
+                bool needRetry;
+                do
+                {
+                    try
+                    {
+                        using var client = _clientFactory.Create(endpoint);
+                        isHealthy = await client.IsServiceHealthy(default);
+                    }
+                    catch (Exception ex)
```

Comment:
when it throws `isHealthy` is not set to `false`, which means it is possible that isHealthy keeps last `true` value, is it by design?

Reply:
The local var `isHealthy` is initialized to `false`. If exception is thrown, `isHealthy` is not set new value and won't keep last true value.